### PR TITLE
Treat discussion's name with no word character as empty

### LINKED ISSUE #4196: Disallow Emoji being used as subject
Currently, when someone creates a discussion with a subject using an emoji (as in the kind used on mobile devices), this results in an inaccessible discussion with a blank subject line. We should either render the emoji or, perhaps simpler, just disallow these non-standard characters from being used in subjects.

### PR DESCRIPTION
**Pull request by DaazKu**
Replace https://github.com/vanilla/vanilla/pull/4390
Fix #4196

### CODE REVIEW

File: applications/vanilla/controllers/class.postcontroller.php [hunk 1]
```diff
@@ -241,13 +241,11 @@ public function discussion($CategoryUrlCode = '') {
                     }
                 }
 
-                // Make sure that the title will not be invisible after rendering
-                if (unicodeRegexSupport()) {
```
Comment (initvector) [inline applications/vanilla/controllers/class.postcontroller.php]: I think we still want to keep the `unicodeRegexSupport` before using UTF regex.

File: applications/vanilla/controllers/class.postcontroller.php [hunk 2]
```diff
@@ -241,13 +241,11 @@ public function discussion($CategoryUrlCode = '') {
...
-                    $Name = preg_replace('/^[\pC ]*(.*?)[\pC ]*$/u', '$1', $this->Form->getFormValue('Name', ''));
-                } else {
-                    $Name = trim($this->Form->getFormValue('Name', ''));
-                }
-                if ($Name != '' && Gdn_Format::text($Name) == '') {
+                $Name = trim($this->Form->getFormValue('Name', ''));
+                // Let's be super aggressive and disallow titles with no word characters in them!
+                $hasWordCharacter = preg_match('/\w/u', $Name) === 1;
+
+                if (!$hasWordCharacter || (!$Name != '' && Gdn_Format::text($Name) == '')) {
```
Comment (initvector) [inline applications/vanilla/controllers/class.postcontroller.php]: Probably don't want the leading `!` on `$Name`.

File: applications/vanilla/views/discussions/table_functions.php [hunk 3]
```diff
@@ -58,7 +58,8 @@ function writeDiscussionRow($Discussion, $Sender, $Session) {
         $Sender->fireEvent('BeforeDiscussionName');
 
         $DiscussionName = $Discussion->Name;
-        if ($DiscussionName == '') {
+        // If there are no word character detected in the title treat it as if it is blank.
+        if (preg_match('/\w/u', $DiscussionName) !== 1) {
```
Comment (initvector) [inline applications/vanilla/views/discussions/table_functions.php]: Is using an is-identical comparison that much faster?  I think pretty much everything in Vanilla that uses `preg_match` currently just tests for the result being truthy/falsey. If it's a real benefit, I think it's worth using, going forward.  As it stands, these comparisons stick out with similar conditionals in core and I want to make sure it's for a good reason.

File: applications/vanilla/controllers/class.postcontroller.php (hunk 2 above)
Comment (DaazKu) [inline applications/vanilla/controllers/class.postcontroller.php]: Indeed

File: applications/vanilla/controllers/class.postcontroller.php (hunk 1 above)
Comment (DaazKu) [inline applications/vanilla/controllers/class.postcontroller.php]: No need. There is no equivalent to /\w/u if unicode is not supported.
And in that case the u modifier will just be ignored.

File: applications/vanilla/views/discussions/table_functions.php (hunk 3 above)
Comment (DaazKu) [inline applications/vanilla/views/discussions/table_functions.php]: Exact match is pretty much always better but I wouldn't call it a real benefit.

**Comment by initvector** [comment]
Seems solid, overall.  I am experiencing one issue:

When I enter a discussion name with emoji and word characters, it passes the test, but saves an empty name in the database.  This leads to a discussion being successfully submitted and saved, but rendering with an empty title.  I believe this is due to Vanilla's encoding for the `GDN_Discussion.Name` field.

**Comment by DaazKu** [comment]
This is indeed a database issue. My guess is that you are using an emojis that needs utf8mb4 instead of utf8.

**Pull request merged**